### PR TITLE
[SPARK-39367][DOCS][SQL] Review and fix issues in Scala/Java API docs of SQL module

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Compare the 3.3.0 API doc with the latest release version 3.2.1. Fix the following issues:

* Add missing Since annotation for new APIs
* Remove the leaking class/object in API doc

### Why are the changes needed?

Improve API docs

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing UT

### CODE REVIEW

**Review by MaxGekk — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 1]
```diff
@@ -46,7 +46,7 @@ import org.apache.spark.sql.types._
  * As commands are executed eagerly, this also includes errors thrown during the execution of
  * commands, which users can see immediately.
  */
-object QueryCompilationErrors extends QueryErrorsBase {
+private[sql] object QueryCompilationErrors extends QueryErrorsBase {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:49]: This was a mistake on 3.2.1.  Let's fix it in 3.3.0

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala [hunk 2]
```diff
@@ -42,7 +42,7 @@ import org.apache.spark.sql.types.{DataType, DoubleType, FloatType}
  * 7. SQL expressions shall be wrapped by double quotes.
  *   For example: "earnings + 1".
  */
-trait QueryErrorsBase {
+private[sql] trait QueryErrorsBase {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryErrorsBase.scala:45]: This was a mistake on 3.2.1.  Let's fix it in 3.3.0

**Review by xuanyuanking — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.3.

**Pull request closed without merging**